library/libwaku.nim

```diff
@@ -0,0 +1,400 @@
+  except JsonParsingError:
+    jsonResp = errResp (fmt"Problem parsing json message. {getCurrentExceptionMsg()}")
+    return false
+
+  var wakuMessage: WakuMessage
+  try:
+    var version = 0'u32
+    if jsonContent.hasKey("version"):
+      version = (uint32) jsonContent["version"].getInt()
+
+    wakuMessage = WakuMessage(
+        # Visit https://rfc.vac.dev/spec/14/ for further details
+        payload: jsonContent["payload"].getStr().toSeq().mapIt(byte (it)),
+        contentTopic: $jsonContent["content_topic"].getStr(),
+        version: version,
+        timestamp: getTime().toUnix(),
+        ephemeral: false
+    )
+  except KeyError:
+    jsonResp = errResp(fmt"Problem building the WakuMessage. {getCurrentExceptionMsg()}")
+    return false
+
+  let targetPubSubTopic = if $pubSubTopic == "":
+                            DefaultPubsubTopic
+                          else:
+                            $pubSubTopic
+
+  if node.wakuRelay.isNil():
+    jsonResp = errResp("Can't publish. WakuRelay is not enabled.")
+    return false
+
+  let pubMsgFut = node.wakuRelay.publish(targetPubSubTopic, wakuMessage)
+
+  # With the next loop we convert an asynchronous call into a synchronous one
+  for i in 0 .. timeoutMs:
+    if pubMsgFut.finished():
+      break
+    sleep(1)
+
+  if pubMsgFut.finished():
+    let numPeers = pubMsgFut.read()
+    if numPeers == 0:
+      jsonResp = errResp("Message not sent because no peers found")
+    elif numPeers > 0:
+      # TODO: pending to return a valid message Id (response when all is correct)
+      jsonResp = okResp("hard-coded-message-id")
+
+  else:
+    jsonResp = errResp("Timeout expired")
+
+  return true
+
+proc waku_start() {.dynlib, exportc.} =
+  waitFor node.start()
+
+proc waku_stop() {.dynlib, exportc.} =
+  waitFor node.stop()
+
+proc waku_relay_subscribe(
+                pubSubTopic: cstring,
+                jsonResp: var string): bool
+                {.dynlib, exportc.} =
+  # @params
+  #  topic: Pubsub topic to subscribe to. If empty, it subscribes to the default pubsub topic.
+  if eventCallback == nil:
+    jsonResp = errResp("""Cannot subscribe without a callback.
+Kindly set it with the 'waku_set_event_callback' function""")
+    return false
+
+  if node.wakuRelay.isNil():
+    jsonResp = errResp("Cannot subscribe without Waku Relay enabled.")
+    return false
+
+  node.wakuRelay.subscribe(PubsubTopic($pubSubTopic), PubsubRawHandler(relayEventCallback))
+
+  jsonResp = okResp("true")
+  return true
+
+proc waku_relay_unsubscribe(
+                pubSubTopic: cstring,
+                jsonResp: var string): bool
+                {.dynlib, exportc.} =
+  # @params
+  #  topic: Pubsub topic to subscribe to. If empty, it unsubscribes to the default pubsub topic.
+  if node.wakuRelay == nil:
+    jsonResp = errResp("""Cannot unsubscribe without a callback.
+Kindly set it with the 'waku_set_event_callback' function""")
+    return false
+
+  if node.wakuRelay.isNil():
```

Comment:
for learning purposes, what's the difference between  `wakuRelay == nil` and `wakuRelay.isNil()`?

Reply:
Good point my friend! Actually you discovered a bug hehe.

It seems `isNil()` is more efficient, so I'd rather use it from now on.

> proc isNil*[T](x: ptr T): bool {.noSideEffect, magic: "IsNil".}
> proc isNil*(x: pointer): bool {.noSideEffect, magic: "IsNil".}
> proc isNil*(x: cstring): bool {.noSideEffect, magic: "IsNil".}
> proc isNil*[T: proc](x: T): bool {.noSideEffect, magic: "IsNil".}
>  Fast check whether `x` is nil. This is sometimes more efficient than ## `== nil`.